market/mysterium/mysterium_api.go

```diff
@@ -236,13 +245,33 @@ func (mApi *MysteriumAPI) QueryProposals(query ProposalsQuery) ([]market.Service
 		return nil, err
 	}
 
-	var proposalsResponse ProposalsResponse
-	err = mApi.http.DoRequestAndParseResponse(req, &proposalsResponse)
+	mApi.latestProposalsMux.Lock()
+	defer mApi.latestProposalsMux.Unlock()
+	req.Header.Add("If-None-Match", mApi.latestProposalsEtag)
+
+	res, err := mApi.http.Do(req)
 	if err != nil {
+		return nil, errors.Wrap(err, "cannot fetch proposals")
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusNotModified {
+		return mApi.latestProposals, nil
+	}
+
+	if err := requests.ParseResponseError(res); err != nil {
 		return nil, err
 	}
+
+	var proposalsResponse ProposalsResponse
+	if err := requests.ParseResponseJSON(res, &proposalsResponse); err != nil {
+		return nil, errors.Wrap(err, "cannot parse proposals response")
+	}
+
+	mApi.latestProposalsEtag = res.Header.Get("ETag")
```

Comment:
I'm not sure about the way you're locking here. You're potentially keeping another routine waiting for 20 seconds(see timeout). The way i'd do it, is extract 2 methods - getLatestEtag and setLatestEtag. Both of those internally would lock the underlying field before getting/setting it. The methods would only be called when needed, preventing excess waiting time for other routines. But otherwise - does look good.